File: typescript/packages/core/src/commands/builtin/generic/nl.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { AsyncLineIterator } from '../../../io/async_line_iterator.ts'
import { IOResult } from '../../../io/types.ts'
import type { PathSpec } from '../../../types.ts'
import type { CommandFnResult, CommandOpts } from '../../config.ts'
import { resolveSource } from '../utils/stream.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

function shouldNumber(line: string, bodyNumbering: string, pattern: RegExp | null): boolean {
  if (bodyNumbering === 'n') return false
  if (bodyNumbering === 'a') return true
  if (bodyNumbering === 'p' && pattern !== null) return pattern.test(line)
  return line.trim() !== ''
}

function padLeft(value: string, width: number): string {
  return value.length >= width ? value : ' '.repeat(width - value.length) + value
}

interface NlOptions {
  bodyNumbering: string
  start: number
  increment: number
  width: number
  separator: string
  pattern: RegExp | null
}

async function* nlStream(
  source: AsyncIterable<Uint8Array>,
  opts: NlOptions,
): AsyncIterable<Uint8Array> {
  let num = opts.start
  const iter = new AsyncLineIterator(source)
  for await (const raw of iter) {
    const line = DEC.decode(raw)
    if (shouldNumber(line, opts.bodyNumbering, opts.pattern)) {
      yield ENC.encode(`${padLeft(String(num), opts.width)}${opts.separator}${line}\n`)
      num += opts.increment
    } else {
      yield ENC.encode(`${' '.repeat(opts.width)}${opts.separator}${line}\n`)
    }
  }
}

async function* nlMulti(
  paths: readonly PathSpec[],
  opts: NlOptions,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
): AsyncIterable<Uint8Array> {
  for (const p of paths) {
    for await (const chunk of nlStream(stream(p), opts)) yield chunk
  }
}

function parseOptions(flags: Record<string, string | boolean>): NlOptions {
  const b = typeof flags.b === 'string' ? flags.b : 't'
  let bodyNumbering = b
  let pattern: RegExp | null = null
  if (b.startsWith('p')) {
    bodyNumbering = 'p'
    pattern = new RegExp(b.slice(1))
  }
  const parseIntFlag = (key: 'v' | 'i' | 'w', fallback: number): number =>
    typeof flags[key] === 'string' ? Number.parseInt(flags[key], 10) : fallback
  return {
    bodyNumbering,
    start: parseIntFlag('v', 1),
    increment: parseIntFlag('i', 1),
    width: parseIntFlag('w', 6),
    separator: typeof flags.s === 'string' ? flags.s : '\t',
    pattern,
  }
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function nlGeneric(
  paths: PathSpec[],
  opts: CommandOpts,
  stream: (p: PathSpec) => AsyncIterable<Uint8Array>,
): Promise<CommandFnResult> {
  const nlOpts = parseOptions(opts.flags)
  if (paths.length > 0) {
    return [nlMulti(paths, nlOpts, stream), new IOResult()]
  }
  try {
    const source = resolveSource(opts.stdin, 'nl: missing operand')
    return [nlStream(source, nlOpts), new IOResult()]
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
  }
}

File: python/mirage/commands/builtin/email/nl.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import re
from collections.abc import AsyncIterator

from mirage.accessor.email import EmailAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.utils.stream import _read_stdin_async
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.email.glob import resolve_glob
from mirage.core.email.read import read as email_read
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _should_number(line: str, body_numbering: str,
                   pattern: re.Pattern[str] | None) -> bool:
    if body_numbering == "n":
        return False
    if body_numbering == "a":
        return True
    if body_numbering == "p" and pattern is not None:
        return pattern.search(line) is not None
    return bool(line.strip())


@command("nl", resource="email", spec=SPECS["nl"])
async def nl(
    accessor: EmailAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    b: str | None = None,
    v: str | None = None,
    i: str | None = None,
    w: str | None = None,
    s: str | None = None,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    body_numbering_raw = b if b is not None else "t"
    pattern: re.Pattern[str] | None = None
    if body_numbering_raw.startswith("p"):
        body_numbering = "p"
        pattern = re.compile(body_numbering_raw[1:])
    else:
        body_numbering = body_numbering_raw
    start = int(v) if v is not None else 1
    increment = int(i) if i is not None else 1
    width = int(w) if w is not None else 6
    separator = s if s is not None else "\t"

    if paths:
        paths = await resolve_glob(accessor, paths, index)
        p = paths[0]
        raw = await email_read(accessor, p, index)
    else:
        raw = await _read_stdin_async(stdin)
        if raw is None:
            raise ValueError("nl: missing operand")

    num = start
    out_lines: list[str] = []
    for line in raw.decode(errors="replace").splitlines():
        if _should_number(line, body_numbering, pattern):
            out_lines.append(f"{num:{width}d}{separator}{line}")
            num += increment
        else:
            out_lines.append(f"{' ' * width}{separator}{line}")
    return ("\n".join(out_lines) + "\n").encode(), IOResult()
